source/core/Array.cpp

```diff
@@ -1351,44 +1351,48 @@ struct ArrayReshapeStruct : public VarArgInstruction
 class ArrayIterator
 {
  private:
-    ArrayDimensionVector  _indexStack;
-    IntIndex _indexDim;
-    IntIndex*  dimensions;
-    IntIndex  _rank;
     TypedArrayCoreRef _array;
+    IntIndex  _rank;
+    IntIndex*  _dimensionLengths;
+    ArrayDimensionVector _indexes;
+
+ private:
+    void ResetIndexes() {
+        for (IntIndex i = 0; i < _rank; i++) {
+            _indexes[i] = 0;
+        }
+    }
+
  public:
     explicit ArrayIterator(TypedArrayCoreRef array) {
         _array = array;
         _rank = array->Rank();
-        _indexDim = 0;
-        dimensions = array->DimensionLengths();
-        for (IntIndex i = 0; i < _rank; i++) {
-            _indexStack[i] = 0;
-        }
+        _dimensionLengths = array->DimensionLengths();
+        ResetIndexes();
     }
+
     void* Begin() {
-        for (IntIndex i = 0; i < _rank; i++) {
-            _indexStack[i] = 0;
-        }
-        _indexDim = 0;
-        return _array->BeginAtND(_rank, _indexStack);
+        ResetIndexes();
+        return _array->BeginAtND(_rank, _indexes);
     }
+
     void* Next() {
-        _indexStack[_indexDim]++;
-        if (dimensions[_indexDim] <= _indexStack[_indexDim]) {
-            while (dimensions[_indexDim] <= _indexStack[_indexDim]) {
-                _indexDim++;
-                _indexStack[_indexDim]++;
-            }
-            if (_indexDim >= _rank) {
-                return NULL;
-            }
-            for (IntIndex i = 0; i < _indexDim; i++) {
-                _indexStack[i] = 0;
-                _indexDim = 0;
+        IntIndex dimensionIndex = 0;
+        while (dimensionIndex < _rank) {
+            _indexes[dimensionIndex]++;
+            if (_indexes[dimensionIndex] >= _dimensionLengths[dimensionIndex]) {
```

Comment:
I see your point. The rigorous check suggests that it is possible in this code for ">" to happen when it actually is not. I prefer a check that is necessary for the algorithm. Something that is more than necessary -- for a defensive style programming -- can be handled via an explicit error or ASSERT in debug builds (whichever is appropriate on a case by case basis)  -- which will again explain the intent of the developer that this MUST not happen. When there is a real need for '>' to happen in the algorithm, then the assert can be removed and the condition can be modified -- which will make  the intent of the code clear. A more rigorous check actually allows for some unintentional code to creep in where the index goes past the length without being equal first and not being caught.